openedx/core/djangoapps/catalog/tests/test_utils.py

```diff
@@ -17,14 +19,108 @@
+        }
+        cache.set_many(partial_programs, None)
+
+        # When called before UUIDs are cached, the function should return an empty
+        # list and log a warning.
+        self.assertEqual(get_programs(), [])
+        mock_warning.assert_called_once_with('Program UUIDs are not cached.')
+        mock_warning.reset_mock()
+
+        # Cache UUIDs for all 3 programs.
+        cache.set(
+            PROGRAM_UUIDS_CACHE_KEY,
+            [program['uuid'] for program in programs],
+            None
+        )
+
+        actual_programs = get_programs()
+
+        # The 2 cached programs should be returned while a warning should be logged
+        # for the missing one.
+        self.assertEqual(
+            set(program['uuid'] for program in actual_programs),
+            set(program['uuid'] for program in partial_programs.values())
+        )
+        mock_warning.assert_called_with(
+            'Details for program {uuid} are not cached.'.format(uuid=programs[2]['uuid'])
+        )
+        mock_warning.reset_mock()
+
+        # We can't use a set comparison here because these values are dictionaries
+        # and aren't hashable. We've already verified that all programs came out
+        # of the cache above, so all we need to do here is verify the accuracy of
+        # the data itself.
+        for program in actual_programs:
+            key = PROGRAM_CACHE_KEY_TPL.format(uuid=program['uuid'])
+            self.assertEqual(program, partial_programs[key])
+
+        # Cache details for all 3 programs.
+        all_programs = {
+            PROGRAM_CACHE_KEY_TPL.format(uuid=program['uuid']): program for program in programs
+        }
+        cache.set_many(all_programs, None)
+
+        actual_programs = get_programs()
+
+        # All 3 programs should be returned.
+        self.assertEqual(
+            set(program['uuid'] for program in actual_programs),
+            set(program['uuid'] for program in all_programs.values())
+        )
+        self.assertFalse(mock_warning.called)
+
+        for program in actual_programs:
+            key = PROGRAM_CACHE_KEY_TPL.format(uuid=program['uuid'])
+            self.assertEqual(program, all_programs[key])
+
+    def test_get_one(self, mock_warning):
+        expected_program = ProgramFactory()
+        expected_uuid = expected_program['uuid']
+
+        self.assertEqual(get_programs(uuid=expected_uuid), None)
+        mock_warning.assert_called_once_with(
+            'Details for program {uuid} are not cached.'.format(uuid=expected_uuid)
+        )
+        mock_warning.reset_mock()
+
+        cache.set(
+            PROGRAM_CACHE_KEY_TPL.format(uuid=expected_uuid),
+            expected_program,
+            None
+        )
+
+        actual_program = get_programs(uuid=expected_uuid)
+        self.assertEqual(actual_program, expected_program)
+        self.assertFalse(mock_warning.called)
+
```

Comment:
should we have a test case that tests what happens if the waffle flag isn't set?